pkg/virt-api/validating-webhook/validating-webhook.go

```diff
@@ -509,6 +509,95 @@ func ValidateVirtualMachineInstanceSpec(field *k8sfield.Path, spec *v1.VirtualMa
 		}
 	}
 
+	// Validate CPU pinning
+	if spec.Domain.CPU != nil && spec.Domain.CPU.DedicatedCPUPlacement {
+		requestsMem := spec.Domain.Resources.Requests.Memory().Value()
+		limitsMem := spec.Domain.Resources.Limits.Memory().Value()
+		requestsCPU := spec.Domain.Resources.Requests.Cpu().Value()
+		limitsCPU := spec.Domain.Resources.Limits.Cpu().Value()
+		vmCores := int64(spec.Domain.CPU.Cores)
+		// memory should be provided
+		if limitsMem == 0 && requestsMem == 0 {
+			causes = append(causes, metav1.StatusCause{
+				Type: metav1.CauseTypeFieldValueInvalid,
+				Message: fmt.Sprintf("%s or %s should be provided",
+					field.Child("domain", "resources", "requests", "memory").String(),
+					field.Child("domain", "resources", "limits", "memory").String(),
+				),
+				Field: field.Child("domain", "resources", "limits", "memory").String(),
+			})
+		}
+
+		// provided CPU requests must be an interger
+		if requestsCPU > 0 && requestsCPU*1000 != spec.Domain.Resources.Requests.Cpu().MilliValue() {
```

Comment:
are we just looking for the mod of the millivalue here to ensure it is divisible by 1000? If so is this logic the same as what I have below?

```
requestsCPU  > 0 && (spec.Domain.Resources.Requests.Cpu().MilliValue() % 1000) > 0
```

Reply:
I did this a first and then looked at what cpu_manager does [1]
I think this is needed to make sure that `.Value()`  didn't round the original value.

[1] https://github.com/NVIDIA/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L201